pkg/virt-controller/watch/vm/vm.go

```diff
@@ -3254,3 +3260,184 @@ func (c *Controller) handleMemoryHotplugRequest(vm *virtv1.VirtualMachine, vmi *
+			(volume.VolumeSource.DataVolume != nil && volume.VolumeSource.DataVolume.Hotpluggable)
+		_, ok := volsVM[volume.Name]
+		if !ok && hotpluggableVol {
+			hotplugOp = true
+		}
+	}
+
+	return hotplugOp
+}
+
+func (c *Controller) patchHotplugVolumes(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+	if !vmi.IsRunning() {
+		return nil
+	}
+
+	newVmiVolumes := append(filterHotplugVMIVolumes(vm, vmi), getNewHotplugVMVolumes(vm, vmi)...)
+	newVmiDisks := append(filterHotplugVMIDisks(vm, vmi, newVmiVolumes), getNewHotplugVMDisks(vm, vmi, newVmiVolumes)...)
+
+	if equality.Semantic.DeepEqual(vmi.Spec.Volumes, newVmiVolumes) &&
+		equality.Semantic.DeepEqual(vmi.Spec.Domain.Devices.Disks, newVmiDisks) {
+		log.Log.Object(vm).V(3).Info("No hotplug volumes to patch")
+		return nil
+	}
+
+	patchSet := patch.New(
+		patch.WithTest("/spec/volumes", vmi.Spec.Volumes),
+		patch.WithTest("/spec/domain/devices/disks", vmi.Spec.Domain.Devices.Disks),
+	)
+
+	if len(vmi.Spec.Volumes) > 0 {
+		patchSet.AddOption(patch.WithReplace("/spec/volumes", newVmiVolumes))
+	} else {
+		patchSet.AddOption(patch.WithAdd("/spec/volumes", newVmiVolumes))
+	}
+
+	if len(vmi.Spec.Domain.Devices.Disks) > 0 {
+		patchSet.AddOption(patch.WithReplace("/spec/domain/devices/disks", newVmiDisks))
+	} else {
+		patchSet.AddOption(patch.WithAdd("/spec/domain/devices/disks", newVmiDisks))
+	}
+
+	patchBytes, err := patchSet.GeneratePayload()
+	if err != nil {
+		return err
+	}
+
+	// XXX DELETEME
+	fmt.Fprint(os.Stderr, "\n"+string(patchBytes)+"\n")
+
+	_, err = c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, patchBytes, v1.PatchOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func filterHotplugVMIVolumes(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) []virtv1.Volume {
+	var volumes []virtv1.Volume
+	vmVolumesByName := storagetypes.GetVolumesByName(&vm.Spec.Template.Spec)
+
+	// remove any volumes missing/changed in the VM spec
+	for _, vmiVolume := range vmi.Spec.Volumes {
+		if storagetypes.IsHotplugVolume(&vmiVolume) {
+			vmVolume, exists := vmVolumesByName[vmiVolume.Name]
+			if !exists {
+				// volume not in VM spec, remove it
+				log.Log.Object(vm).Infof("Removing hotplug volume %s from VMI, no longer in VM", vmiVolume.Name)
+				continue
+			}
+
+			// volume changed in VM spec - remove it to be re-added with new values later
+			if storagetypes.IsHotplugVolume(vmVolume) && !equality.Semantic.DeepEqual(vmVolume, &vmiVolume) {
+				log.Log.Object(vm).Infof("Removing hotplug volume %s from VMI, volume changed", vmiVolume.Name)
+				continue
+			}
+		}
+
+		volumes = append(volumes, *vmiVolume.DeepCopy())
+	}
+
+	return volumes
+}
+
+func volumesByName(volumes []virtv1.Volume) map[string]*virtv1.Volume {
+	volumeMap := make(map[string]*virtv1.Volume)
+	for _, v := range volumes {
+		volumeMap[v.Name] = v.DeepCopy()
```

Comment:
Why the DeepCopy here? it doesn't look like we are mutating the volumes? Maybe I am missing where we are mutating them.

Reply:
To me, I think it is good practice to deep copy. Maybe the volumes aren't mutated now but who knows about the future? A little extra memory now could save us from a bug in the future. We do deepcopy in similar functions as well:

https://github.com/mhenriks/kubevirt/blob/6844567b8c03b57506abb27a0d4b045562b969c7/pkg/storage/types/pvc.go#L281-L295

Reply:
I'm also in favor of deepcopies! Can we use the function `GetVolumesByName`  from the storage types package?